### ISSUE TITLE
Build fails with Linux kernel 5.3

### ISSUE THREAD

**Issue by ottok** [ottok/shannon-direct-io-pcie-driver#4]
Seems build fails with Linux kernel 5.3. I wonder if the driver has been tested with recent kernels?

```
DKMS make.log for shannon-3.4.0 for kernel 5.3.0-28-generic (x86_64)
to 30.1.2020 05.52.07 +0200
make \
    -C /lib/modules/5.3.0-28-generic/build \
    SHANNON_DRIVER_DIR=/var/lib/dkms/shannon/3.4.0/build \
    SUBDIRS=/var/lib/dkms/shannon/3.4.0/build \
    CONFIG_BLK_DEV_SHANNON=m \
    CONFIG_SHANNON_EMU= \
    EXTRA_CFLAGS=" -DSHANNON_RELEASE" \
    INSTALL_MOD_PATH= \
    modules
make[1]: Entering directory '/usr/src/linux-headers-5.3.0-28-generic'
Makefile:227: ================= WARNING ================
Makefile:228: 'SUBDIRS' will be removed after Linux 5.3
Makefile:229: 
Makefile:230: If you are building an individual subdirectory
Makefile:231: in the kernel tree, you can do like this:
Makefile:232: $ make path/to/dir/you/want/to/build/
Makefile:233: (Do not forget the trailing slash)
Makefile:234: 
Makefile:235: If you are building an external module,
Makefile:236: Please use 'M=' or 'KBUILD_EXTMOD' instead
Makefile:237: ==========================================
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_main.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_ftl.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_prefetch.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_boot.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_epilog.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_err_handler.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_ns.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_ioctl.o
  SHIPPED /var/lib/dkms/shannon/3.4.0/build/shannon_scsi_cmd.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_kcore.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_time.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_workqueue.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_waitqueue.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_file.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_pci.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_scatter.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_device.o
  CC [M]  /var/lib/dkms/shannon/3.4.0/build/shannon_block.o
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c: In function ‘shannon_end_request’:
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1021:2: error: implicit declaration of function ‘blk_end_request’; did you mean ‘blk_get_request’? [-Werror=implicit-function-declaration]
  blk_end_request(rq, error ? 0 : -EIO, blk_rq_bytes(rq));
  ^~~~~~~~~~~~~~~
  blk_get_request
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c: In function ‘shannon_disk_request’:
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1397:15: error: implicit declaration of function ‘blk_peek_request’; did you mean ‘blk_get_request’? [-Werror=implicit-function-declaration]
  while ((rq = blk_peek_request(q))) {
               ^~~~~~~~~~~~~~~~
               blk_get_request
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1397:13: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  while ((rq = blk_peek_request(q))) {
             ^
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1398:3: error: implicit declaration of function ‘blk_start_request’; did you mean ‘blk_abort_request’? [-Werror=implicit-function-declaration]
   blk_start_request(rq);
   ^~~~~~~~~~~~~~~~~
   blk_abort_request
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1403:19: error: incompatible type for argument 1 of ‘spin_unlock_irq’
   spin_unlock_irq(q->queue_lock);
                   ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1:
./include/linux/spinlock.h:386:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_unlock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1407:18: error: incompatible type for argument 1 of ‘spin_lock_irq’
    spin_lock_irq(q->queue_lock);
                  ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1:
./include/linux/spinlock.h:361:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_lock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1408:4: error: implicit declaration of function ‘blk_requeue_request’; did you mean ‘blk_update_request’? [-Werror=implicit-function-declaration]
    blk_requeue_request(q, rq);
    ^~~~~~~~~~~~~~~~~~~
    blk_update_request
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1409:20: error: incompatible type for argument 1 of ‘spin_unlock_irq’
    spin_unlock_irq(q->queue_lock);
                    ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1:
./include/linux/spinlock.h:386:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_unlock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1411:17: error: incompatible type for argument 1 of ‘spin_lock_irq’
   spin_lock_irq(q->queue_lock);
                 ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1:
./include/linux/spinlock.h:361:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_lock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c: In function ‘shannon_init_queue’:
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1438:10: error: implicit declaration of function ‘blk_init_queue’; did you mean ‘blk_put_queue’? [-Werror=implicit-function-declaration]
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
          ^~~~~~~~~~~~~~
          blk_put_queue
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1438:26: error: ‘request_fn_proc’ undeclared (first use in this function); did you mean ‘request_nmi’?
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
                          ^~~~~~~~~~~~~~~
                          request_nmi
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1438:26: note: each undeclared identifier is reported only once for each function it appears in
/var/lib/dkms/shannon/3.4.0/build/shannon_block.c:1438:43: error: expected expression before ‘)’ token
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
                                           ^
cc1: some warnings being treated as errors
scripts/Makefile.build:288: recipe for target '/var/lib/dkms/shannon/3.4.0/build/shannon_block.o' failed
make[2]: *** [/var/lib/dkms/shannon/3.4.0/build/shannon_block.o] Error 1
Makefile:1655: recipe for target '_module_/var/lib/dkms/shannon/3.4.0/build' failed
make[1]: *** [_module_/var/lib/dkms/shannon/3.4.0/build] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.3.0-28-generic'
Makefile:41: recipe for target 'modules' failed
make: *** [modules] Error 2
```

**Comment by ottok** [comment]
Tested with version 3.4.1 and still the same error:
```
/var/lib/dkms/shannon/3.4.1/build$ cat make.log 
DKMS make.log for shannon-3.4.1 for kernel 5.3.0-62-generic (x86_64)
to 23.7.2020 01.17.17 +0300
make \
    -C /lib/modules/5.3.0-62-generic/build \
    SHANNON_DRIVER_DIR=/var/lib/dkms/shannon/3.4.1/build \
    SUBDIRS=/var/lib/dkms/shannon/3.4.1/build \
    CONFIG_BLK_DEV_SHANNON=m \
    CONFIG_SHANNON_EMU= \
    EXTRA_CFLAGS=" -DSHANNON_RELEASE" \
    INSTALL_MOD_PATH= \
    modules
make[1]: Entering directory '/usr/src/linux-headers-5.3.0-62-generic'
Makefile:227: ================= WARNING ================
Makefile:228: 'SUBDIRS' will be removed after Linux 5.3
Makefile:229: 
Makefile:230: If you are building an individual subdirectory
Makefile:231: in the kernel tree, you can do like this:
Makefile:232: $ make path/to/dir/you/want/to/build/
Makefile:233: (Do not forget the trailing slash)
Makefile:234: 
Makefile:235: If you are building an external module,
Makefile:236: Please use 'M=' or 'KBUILD_EXTMOD' instead
Makefile:237: ==========================================
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_main.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_ftl.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_prefetch.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_boot.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_epilog.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_err_handler.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_ns.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_ioctl.o
  SHIPPED /var/lib/dkms/shannon/3.4.1/build/shannon_scsi_cmd.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_kcore.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_time.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_workqueue.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_waitqueue.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_file.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_pci.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_scatter.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_device.o
  CC [M]  /var/lib/dkms/shannon/3.4.1/build/shannon_block.o
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c: In function ‘shannon_end_request’:
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1021:2: error: implicit declaration of function ‘blk_end_request’; did you mean ‘blk_get_request’? [-Werror=implicit-function-declaration]
  blk_end_request(rq, error ? 0 : -EIO, blk_rq_bytes(rq));
  ^~~~~~~~~~~~~~~
  blk_get_request
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c: In function ‘shannon_disk_request’:
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1397:15: error: implicit declaration of function ‘blk_peek_request’; did you mean ‘blk_get_request’? [-Werror=implicit-function-declaration]
  while ((rq = blk_peek_request(q))) {
               ^~~~~~~~~~~~~~~~
               blk_get_request
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1397:13: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  while ((rq = blk_peek_request(q))) {
             ^
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1398:3: error: implicit declaration of function ‘blk_start_request’; did you mean ‘blk_abort_request’? [-Werror=implicit-function-declaration]
   blk_start_request(rq);
   ^~~~~~~~~~~~~~~~~
   blk_abort_request
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1403:19: error: incompatible type for argument 1 of ‘spin_unlock_irq’
   spin_unlock_irq(q->queue_lock);
                   ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1:
./include/linux/spinlock.h:386:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_unlock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1407:18: error: incompatible type for argument 1 of ‘spin_lock_irq’
    spin_lock_irq(q->queue_lock);
                  ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1:
./include/linux/spinlock.h:361:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_lock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1408:4: error: implicit declaration of function ‘blk_requeue_request’; did you mean ‘blk_update_request’? [-Werror=implicit-function-declaration]
    blk_requeue_request(q, rq);
    ^~~~~~~~~~~~~~~~~~~
    blk_update_request
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1409:20: error: incompatible type for argument 1 of ‘spin_unlock_irq’
    spin_unlock_irq(q->queue_lock);
                    ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1:
./include/linux/spinlock.h:386:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_unlock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1411:17: error: incompatible type for argument 1 of ‘spin_lock_irq’
   spin_lock_irq(q->queue_lock);
                 ^
In file included from ./include/linux/wait.h:9:0,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from /var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1:
./include/linux/spinlock.h:361:29: note: expected ‘spinlock_t * {aka struct spinlock *}’ but argument is of type ‘spinlock_t {aka struct spinlock}’
 static __always_inline void spin_lock_irq(spinlock_t *lock)
                             ^~~~~~~~~~~~~
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c: In function ‘shannon_init_queue’:
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1438:10: error: implicit declaration of function ‘blk_init_queue’; did you mean ‘blk_put_queue’? [-Werror=implicit-function-declaration]
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
          ^~~~~~~~~~~~~~
          blk_put_queue
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1438:26: error: ‘request_fn_proc’ undeclared (first use in this function); did you mean ‘request_nmi’?
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
                          ^~~~~~~~~~~~~~~
                          request_nmi
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1438:26: note: each undeclared identifier is reported only once for each function it appears in
/var/lib/dkms/shannon/3.4.1/build/shannon_block.c:1438:43: error: expected expression before ‘)’ token
  queue = blk_init_queue((request_fn_proc *)shannon_disk_request,
                                           ^
cc1: some warnings being treated as errors
scripts/Makefile.build:288: recipe for target '/var/lib/dkms/shannon/3.4.1/build/shannon_block.o' failed
make[2]: *** [/var/lib/dkms/shannon/3.4.1/build/shannon_block.o] Error 1
Makefile:1664: recipe for target '_module_/var/lib/dkms/shannon/3.4.1/build' failed
make[1]: *** [_module_/var/lib/dkms/shannon/3.4.1/build] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.3.0-62-generic'
Makefile:41: recipe for target 'modules' failed
make: *** [modules] Error 2
```

**Comment by ottok** [comment]
Had issue with paths. This is now fixed and 3.4.1 works!

**Comment by ottok** [comment]
Had issue with paths. This is now fixed and 3.4.1 works!
